DBAgenda.php

```diff
@@ -0,0 +1,311 @@
+            }
+        }
+        
+        $query = $this->pdo->prepare("REPLACE INTO events (vCalendarFilename, ETag, datetime_begin, number_volunteers_required, vCalendarRaw) VALUES (:vCalendarFilename, :ETag, :datetime_begin, :number_volunteers_required, :vCalendarRaw)");
+        $query->execute(array(
+            'vCalendarFilename' =>  $event['vCalendarFilename'],
+            'ETag' => $event['ETag'],
+            'datetime_begin' => $datetime_begin->format('Y-m-d H:i:s'),
+            'number_volunteers_required' => $number_volunteers_required,
+            'vCalendarRaw' => $event['vCalendarRaw']
+        ));
+        
+        if(isset($vCalendar->VEVENT->ATTENDEE)) {
+            foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                $mail = str_replace("mailto:", "", (string)$attendee);
+
+                $query = $this->pdo->prepare("SELECT * FROM attendees WHERE email=:email;");
+                $query->execute(array(
+                    'email' => $mail
+                ));
+
+                if(is_array($ret = $query->fetch())) {
+                    $this->log->debug("attendee: $mail already exists.");
+                } else {
+                    $user = $this->api->users_lookupByEmail($mail);
+                    if(!is_null($user)) {
+                        $userid = $user->id;
+                    } else {
+                        $userid = null;
+                    }
+
+                    $query = $this->pdo->prepare("REPLACE INTO attendees (email, userid) VALUES (:email, :userid)");
+                    $query->execute(array(
+                        'email' =>  $mail,
+                        'userid' =>  $userid
+                    ));
+                }
+                
+                $query = $this->pdo->prepare("INSERT INTO events_attendees (vCalendarFilename, email) VALUES (:vCalendarFilename, :email)");
+                $query->execute(array(
+                    'vCalendarFilename' =>  $event['vCalendarFilename'],
+                    'email' =>  $mail
+                ));
+            }
+        }
+        
+        if(isset($vCalendar->VEVENT->CATEGORIES)) {
+            foreach($vCalendar->VEVENT->CATEGORIES as $category) {
+                $category = (string)$category;
+                
+                if(is_number_of_attendee_category($category)) {
+                    continue;
+                }
+                
+                $query = $this->pdo->prepare("SELECT * FROM categories WHERE name=:name;");
+                $query->execute(array(
+                    'name' => $category
+                ));
+                
+                $id = null;
+                if(is_array($ret = $query->fetch())) {
+                    $id = $ret['id'];
+                    $this->log->debug("category: $category already exists.");
+                } else {                    
+                    $query = $this->pdo->prepare("INSERT INTO categories (name) VALUES (:name);");
```

Comment:
Actually I think we may have a race condition here. Assume this scenario:
- an event is created on the caldav server with a new categories
- two users open the app at the same time
- the code for both users run the SELECT on L248 at the almost same time, so both consider that the new category does not exist
- then the code for both users run this INSERT at the almost same time

then this INSERT will succeed only for the first user and fail for the 2nd one.

I think a convenient way to handle this could be:
- ensure that this line behaves well if the INSERT well because of the UNIQUE constraint (ie: ensure it doesn't end up throwing an exception; if it does then catch it silently)
- instead of retrieving the id of the last inserted row on L263, redo the SELECT as on L248
- completely remove the method `getLastInsertedRowId` because I think we wouldn't need it then

Reply:
The exception to catch is: `Fatal error: Uncaught PDOException: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'InterventionDPE' for key 'name'`. Fixed in f4d2242.

- https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id;
"The ID that was generated is maintained in the server on a per-connection basis. This means that the value returned by the function to a given client is the first AUTO_INCREMENT value generated for most recent statement affecting an AUTO_INCREMENT column by that client. This value cannot be affected by other clients, even if they generate AUTO_INCREMENT values of their own. This behavior ensures that each client can retrieve its own ID without concern for the activity of other clients, and without the need for locks or transactions."

At least we are safe with MySQL (I need to check that for sqlite also).

Reply:
- https://www.sqlite.org/c3ref/last_insert_rowid.html;
The sqlite3_last_insert_rowid(**D**) interface usually returns the rowid of the most recent successful INSERT into a rowid table or virtual table on **database connection D**. Inserts into WITHOUT ROWID tables are not recorded. If no successful INSERTs into rowid tables have ever occurred on the database connection D, then sqlite3_last_insert_rowid(D) returns zero.

So I guess we are safe, it is connection dependent.
Moreover, I have not notice it before, but there is `PDO::lastInsertId` method. Maybe we could use it instead of `getLastInsertedRowId`.